### PR TITLE
Re-factor Issuer Validator to Follow New Validation Model

### PR DESCRIPTION
**Pull request by FuPingFranco**
# Re-factor Issuer Validator to Follow New Validation Model

## Description

- Removed checks for `ValidIssuer` from Issuer validation.
- Moved away from using `TokenValidationParameters` in favor of the new `ValidationParameters`.
- Added new tests in ValidationParameters class.
- Fixed minor inconsistencies on the code.

### CODE REVIEW

File: src/Microsoft.IdentityModel.Tokens/Validation/Validators.Issuer.cs [hunk 1]
```diff
@@ -95,10 +91,8 @@ public static partial class Validators
             if (validationParameters.ConfigurationManager != null)
                 configuration = await validationParameters.ConfigurationManager.GetBaseConfigurationAsync(cancellationToken).ConfigureAwait(false);
 
-            // Throw if all possible places to validate against are null or empty
-            if (string.IsNullOrWhiteSpace(validationParameters.ValidIssuer)
-                && validationParameters.ValidIssuers.IsNullOrEmpty()
-                && string.IsNullOrWhiteSpace(configuration?.Issuer))
+            // Return failed IssuerValidationResult if all possible places to validate against are null or empty.
+            if (validationParameters.ValidIssuers.IsNullOrEmpty() && string.IsNullOrWhiteSpace(configuration?.Issuer))
```
Comment (iNinja) [inline src/Microsoft.IdentityModel.Tokens/Validation/Validators.Issuer.cs]: Similar to Brent's comment on the other PR, now that `ValidIssuers` is an `IList`, we should check `.Count == 0`.

**Review by iNinja — DISMISSED** [review]
Minor change and we're good.

File: src/Microsoft.IdentityModel.Tokens/Validation/ValidationParameters.cs [hunk 2]
```diff
@@ -21,6 +20,7 @@ internal class ValidationParameters
         private string _nameClaimType = ClaimsIdentity.DefaultNameClaimType;
         private string _roleClaimType = ClaimsIdentity.DefaultRoleClaimType;
         private Dictionary<string, object> _instancePropertyBag;
+        private IList<string> _validIssuers = [];
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/Validation/ValidationParameters.cs]: Leave these _validIssuers and _validTokenTypes null.
Use the pattern shown in the Get below.

In general, we should never have setters on properties that are ICollection<T> or IList<T> so we never have to check for null and can just use count.

File: src/Microsoft.IdentityModel.Tokens/Validation/ValidationParameters.cs [hunk 3]
```diff
@@ -518,10 +518,16 @@ public TypeValidatorDelegate TypeValidator
...
-        /// The default is <c>null</c>.
+        /// Gets or sets the <see cref="IList{String}"/> that contains valid issuers that will be used to check against the token's issuer.
+        /// The default is an empty collection.
         /// </summary>
-        public IList<string> ValidIssuers { get; }
+        /// <exception cref="ArgumentNullException">Thrown when the value is set as null.</exception>
+        /// <returns>The <see cref="IList{String}"/> that contains valid issuers that will be used to check against the token's 'iss' claim.</returns>
+        public IList<string> ValidIssuers
+        {
+            get { return _validIssuers; }
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/Validation/ValidationParameters.cs]: We should remove the set and use this patern: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/149148109446c4aa6a341cb4d766ec27b5213bdd/src/Microsoft.IdentityModel.Tokens/SecurityTokenDescriptor.cs#L28

**Review by brentschmaltz — APPROVED** [review]

**Review by brentschmaltz — DISMISSED** [review]
Properties that return IList<T> should not be settable.
If the property can be null, for a successful validation, use a delayed creation.

File: src/Microsoft.IdentityModel.Tokens/Validation/Validators.Issuer.cs (hunk 1 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens/Validation/Validators.Issuer.cs]: Looks like we are still using `IsNullOrEmpty()` on an `IList`.

The comment above needs to be addressed.

File: src/Microsoft.IdentityModel.Tokens/Validation/ValidationParameters.cs (hunk 3 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Tokens/Validation/ValidationParameters.cs]: Updated

**Review by iNinja — APPROVED** [review]

**Review by SaurabhMSFT — APPROVED** [review]

**Pull request merged**